hilti/toolchain/include/ast/node.h

```diff
@@ -483,6 +504,260 @@ class None : public NodeBase, public util::type_erasure::trait::Singleton {
+    using reference = BaseIterator::reference;
+    using iterator_category = BaseIterator::iterator_category;
+
+    explicit RangeIterator(BaseIterator i) : _iter(std::move(i)) {}
+    RangeIterator(const RangeIterator& other) = default;
+    RangeIterator(RangeIterator&& other) = default;
+    RangeIterator() {}
+    ~RangeIterator() = default;
+
+    const Node& node() const { return *_iter; }
+
+    RangeIterator& operator=(const RangeIterator& other) = default;
+    RangeIterator& operator=(RangeIterator&& other) = default;
+    const T& operator*() const { return value(); }
+    const T* operator->() const { return &value(); }
+    bool operator==(const RangeIterator& other) const { return _iter == other._iter; }
+    bool operator!=(const RangeIterator& other) const { return ! (*this == other); }
+
+    RangeIterator operator++(int) {
+        auto x = RangeIterator(_iter);
+        ++_iter;
+        return x;
+    }
+
+    RangeIterator& operator++() {
+        ++_iter;
+        return *this;
+    }
+
+    RangeIterator& operator+=(difference_type i) {
+        _iter += i;
+        return *this;
+    }
+
+    RangeIterator& operator-=(difference_type i) {
+        _iter -= i;
+        return *this;
+    }
+
+    difference_type operator-(const RangeIterator& other) const { return _iter - other._iter; }
+    RangeIterator operator-(difference_type i) const { return RangeIterator(_iter - i); }
+    RangeIterator operator+(difference_type i) const { return RangeIterator(_iter + i); }
+
+private:
+    const T& value() const { return (*_iter).template as<std::remove_const_t<T>>(); }
+
+    BaseIterator _iter;
+};
+
+/**
+ * A range of AST nodes, defined by start and end into an existing vector of
+ * nodes. The range creates a view that can be iterated over, yielding a
+ * reference to each node in turn.
+ */
+template<typename T>
+class Range {
+public:
+    using iterator = RangeIterator<T>;
+    using const_iterator = RangeIterator<T>;
+
+    explicit Range() {}
+    Range(std::vector<Node>::const_iterator begin, std::vector<Node>::const_iterator end)
+        : _begin(std::move(begin)), _end(std::move(end)) {}
+
+    explicit Range(const std::vector<Node>& nodes) : Range(nodes.begin(), nodes.end()) {}
+
+    Range(const Range& other) = default;
+    Range(Range&& other) = default;
+    ~Range() = default;
+
+    auto begin() const { return const_iterator(_begin); }
+    auto end() const { return const_iterator(_end); }
+    size_t size() const { return static_cast<size_t>(_end - _begin); }
+    const T& front() const { return *_begin; }
+    bool empty() const { return _begin == _end; }
+
+    /**
+     * Returns a new vector containing copies of all nodes that the range
+     * includes.
+     **/
+    std::vector<T> copy() const {
+        std::vector<T> x;
+        for ( auto i = _begin; i != _end; i++ )
+            x.push_back(*i);
+
+        return x;
+    }
+
+    const T& operator[](size_t i) const { return *(_begin + i); }
```

Comment:
Let's add some form of check here that `i< std::distance(_begin, _end` (e.g., always check, or just debug mode assertion). I just ran into a situation where after (probably incorrect) optimizations `i` pointed to a non-existing element.

Reply:
Adding an assert.